lib/Hakyll/Web/Html/Compress.hs

```diff
@@ -0,0 +1,101 @@
+-- and it works as follows:
+--
+-- * Comments are removed.
+-- * Several consecutive whitespaces are replaced by a single one, unless within a <pre> tag.
+-- * Within a <pre> tag, @n@ consecutive whitespaces are replaced by a single @\t@ character.
+--   This is useful if a page is heavy on code listings.
+--   Don't forget to add @tab-size: n@ to your CSS!
+--
+-- Any of these steps can be disabled, see 'CompressHtmlOpts'.
+
+module Hakyll.Web.Html.Compress
+  ( CompressHtmlOpts(..)
+  , def
+  , compressHtml
+  , compressHtmlCompiler
+  ) where
+
+import           Data.Char
+import           Data.Default
+import qualified Data.Set as S
+import           Hakyll.Core.Compiler
+import           Hakyll.Core.Item
+import           Hakyll.Web.Html
+import           Text.HTML.TagSoup
+
+-- | The configuration for the HTML compression.
+data CompressHtmlOpts = CompressHtmlOpts
+  { choRemoveComments :: Bool          -- ^ Whether to remove comments.
+  , choCompressWhitespace :: Bool      -- ^ Whether to remove excessive whitespaces.
+  , choTabSize :: Maybe Int            -- ^ Replace this many spaces in <pre> with @\t@ (if 'choCompressWhitespace' is set).
+  }
+
+instance Default CompressHtmlOpts where
+  def = CompressHtmlOpts
+          { choRemoveComments = True
+          , choCompressWhitespace = True
+          , choTabSize = Nothing
+          }
+
+-- | Compiler form of 'compressHtml'.
+compressHtmlCompiler :: CompressHtmlOpts -> Item String -> Compiler (Item String)
+compressHtmlCompiler opts item = pure $ compressHtml opts <$> item
+
+-- | Compresses an HTML string according to the given configuration.
+compressHtml :: CompressHtmlOpts -> String -> String
+compressHtml CompressHtmlOpts{ .. } = withTagList go
+  where
+    go = foldr (.) id
+        $ [ f
+          | (True, f) <- [ (choRemoveComments, removeComments)
+                         , (choCompressWhitespace, compressWS choTabSize)
+                         ]
+          ]
+
+removeComments :: [Tag String] -> [Tag String]
+removeComments = filter (not . isTagComment)
+
+compressWS :: Maybe Int -> [Tag String] -> [Tag String]
+compressWS maybeTabSize = go mempty
+  where
+    go stack =
+      \case [] -> []
+            (tag@(TagClose n) : rest) -> tag : go (S.delete n stack) rest
+            (tag@(TagOpen n _) : rest) -> tag : go (S.insert n stack) rest
+            (tag@(TagText text) : rest)
+              -- all spaces within a <pre> are important, but we can replace them with tabs
+              | "pre" `S.member` stack -> case maybeTabSize of
+                                               Nothing -> tag : go stack rest
+                                               Just tabSize -> TagText (collapseIntoTabs (tabSize - 1) text) : go stack rest
+              | otherwise -> let text' = collapseSpaces text
+                              in case text' of
+                                      [] -> go stack rest
```

Comment:
`collapseSpaces` would only return an empty string if the input was empty. What if the author meant it that way though, e.g. as a marker to later insert something with JS? If there is no rationale for removing tags here, I'd prefer to keep them.

Reply:
This only removes the text node, keeping the surrounding tags intact. Tagsoup [requires](https://hackage.haskell.org/package/tagsoup-0.14.8/docs/Text-HTML-TagSoup.html#t:Tag) `TagText` to contain a non-empty string, and while right now, indeed, `collapseSpaces` returns an empty string only on empty inputs, this might change in the future shall anybody change its logic. I think it's easier to add a safeguard against that.

Reply:
Ah, thanks for the explanation! Please disregard this review comment then.